flask_dance/contrib/azure.py

```diff
@@ -76,6 +79,15 @@ def make_azure_blueprint(
     :returns: A :doc:`blueprint <flask:blueprints>` to attach to your Flask app.
     """
     scope = scope or ["openid", "email", "profile", "User.Read"]
+    token_url = f"https://login.microsoftonline.com/{tenant}/oauth2/v2.0/token"
+    authorization_url = (
+        f"https://login.microsoftonline.com/{tenant}/oauth2/v2.0/authorize"
+    )
+    auto_refresh_url = None  # see #161
+    if offline:
+        scope = [scope] if not isinstance(scope, list) else scope
+        scope.append("offline_access")
```

Comment:
Modifying the `scope` variable feels like a bad pattern to me. If a developer uses `flask-dance` and specifies the `scope` variable, they expect that the contents of that variable are passed to the OAuth provider as they specified it; I can see this lead to many "WTFs" when debugging.

Assuming that the `offline_access` scope is required to support the token refresh feature, I think it's better to invert the logic here. Rather than this function accepting an `offline` boolean and modifying the scope, this function could not add a new variable, and simply check to see if `"offine_access"` is present in the `scope` variable. If so, the token refresh feature is automatically enabled; if not, then it is not.

What do you think?

Reply:
Conceptually, I think I like that idea.  I guess it depends on which weights the most,

1) A common interface, and abstract away technical details
2) A light-weight (technically) transparent interface

If we choose (1), we should stick with the `offline`  keyword (or a similar syntax) as it defines a common way to enable offline access mode across contribs. We thus abstract away the fact that Azure uses a `scope` modification, while e.g. Google uses a `authorization_url_params` modification.

If we choose (2), I believe your proposed syntax is preferred. If case we go down this route, we may consider changing other contribs (e.g. Google) accordingly.

One thing to note though is that even though the scope of the session is modified, the scope of the resulting token is not,

![image](https://github.com/singingwolfboy/flask-dance/assets/1623542/35b44a6c-9fa2-4fc1-81be-22d4f3866412)

I am not sure if this affects your suggestion.

Reply:
> Conceptually, I think I like that idea. I guess it depends on which weights the most,
> 
> 1. A common interface, and abstract away technical details
> 2. A light-weight (technically) transparent interface

Unfortunately, I think that option 1 is not possible. The only common interface between different OAuth providers is OAuth itself -- and more often than not, providers are not even 100% compliant with the OAuth spec! Trying to shoehorn different providers into a common interface is not going to work. Just because Google does it one way, doesn't mean that Microsoft is going to follow suit; in fact, competing companies often try to be intentionally different from each other!

I think making the interface clear & debuggable is a better approach. How would you suggest changing the Google contrib to make it more clear & debuggable?

Also, I was not aware that Azure returns a token with different scopes than the requested scopes -- but if anything, it just provides more evidence that trying to provide a unified interface across providers is not possible. We should not try to obscure these implementation details, particularly since the provider could change their implementation details at any moment.

Reply:
That's true. While a coherent wrapping would be nice, keeping the library maintained to accomodate the latest changes across contribs is infeasible. I'll change the implementation to match your suggestion.